xbmc/pvr/PVRManager.cpp

```diff
@@ -758,45 +759,245 @@ void CPVRManager::ResetPlayingTag(void)
 
 bool CPVRManager::ToggleRecordingOnChannel(unsigned int iChannelId)
 {
-  bool bReturn = false;
-
-  CPVRChannelPtr channel(m_channelGroups->GetChannelById(iChannelId));
+  const CPVRChannelPtr channel(m_channelGroups->GetChannelById(iChannelId));
   if (!channel)
-    return bReturn;
+    return false;
 
-  if (m_addons->HasTimerSupport(channel->ClientID()))
+  return SetRecordingOnChannel(channel, !channel->IsRecording());
+}
+
+void CPVRManager::StartRecordingOnPlayingChannel(bool bOnOff)
+{
+  // can be called from VideoPlayer thread. SetRecordingOnChannel can open a dialog. Thus, execute async.
+  CJobManager::GetInstance().AddJob(new CPVRSetRecordingOnChannelJob(m_addons->GetPlayingChannel(), bOnOff), NULL);
+}
+
+namespace
+{
+enum PVRRECORD_INSTANTRECORDACTION
+{
+  PVRRECORD_INSTANTRECORDACTION_NONE = -1,
```

Comment:
Fair enough. Let's see whether this works without creating any preprocessor problems.